### PR TITLE
Implement JsonTypeInfoResolver.WithAddedModifier

### LINKED ISSUE #86440: Add a `JsonTypeInfoResolver.WithModifier` extension method
### Background and motivation

When contract customization was launched in .NET 7, we included a `Modifiers` property in the `DefaultJsonTypeInfoResolver` class which let users introduce tweaks to the reflection-based contract:

```C#
var options = new JsonSerializerOptions
{
    TypeInfoResolver = new DefaultJsonTypeInfoResolver
    {
        Modifiers = 
        {
           static typeInfo =>
           {
                foreach (JsonPropertyInfo prop in typeInfo.Properties)
                {
                    prop.Name = prop.PropertyType.Name.ToUpperInvariant();
                }
           }
        }
    }
};
```

This particular pattern is not exclusive to reflection-based serialization, but can be applied to the `IJsonTypeInfoResolver` abstraction in the general case. At this point however, we have no baked-in API that lets users do this easily.

### API Proposal

```diff
namespace System.Text.Json.Serialization.Metadata;

public static class JsonTypeInfoResolver
{
    public static IJsonTypeInfoResolver Combine(params IJsonTypeInfoResolver?[] resolvers);
+   public static IJsonTypeInfoResolver WithAddedModifier(this IJsonTypeInfoResolver resolver, Action<JsonTypeInfo> modifier);
}
```

### API Usage

Users can now specify modifiers on top of source generated metadata:
```C#
var options = new JsonSerializerOptions
{
    TypeInfoResolver = SourceGenerationContext.Default
        .WithModifier(
            static typeInfo =>
            {
                foreach (JsonPropertyInfo prop in typeInfo.Properties)
                {
                    prop.Name = prop.PropertyType.Name.ToUpperInvariant();
                }
            })
};
```

### Alternative Designs

The method could be made to accept a param array of modifiers:
```diff
namespace System.Text.Json.Serialization.Metadata;

public static class JsonTypeInfoResolver
{
    public static IJsonTypeInfoResolver Combine(params IJsonTypeInfoResolver?[] resolvers);
+   public static IJsonTypeInfoResolver WithModifiers(this IJsonTypeInfoResolver resolver, params Action<JsonTypeInfo>[] modifiers);
}
```
But I figured it might be preferable if multiple callbacks are chained using fluent-style chaining:
```C#
resolver.WithModifier(modifier1).WithModifier(modifier2);
```

### Risks

_No response_

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Fix #86440.

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoResolverWithAddedModifiers.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace System.Text.Json.Serialization.Metadata
+{
+    internal sealed class JsonTypeInfoResolverWithAddedModifiers : IJsonTypeInfoResolver
+    {
+        private readonly IJsonTypeInfoResolver _source;
+        private readonly Action<JsonTypeInfo>[] _modifiers;
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoResolverWithAddedModifiers.cs:11]: why not List?

Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoResolverWithAddedModifiers.cs:11]: I'm curious if you could support it by List and count (you always append so technically you could reuse same instance) or ImmutableArray/List

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoResolverWithAddedModifiers.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+        public JsonTypeInfoResolverWithAddedModifiers(IJsonTypeInfoResolver source, Action<JsonTypeInfo>[] modifiers)
+        {
+            Debug.Assert(modifiers.Length > 0);
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoResolverWithAddedModifiers.cs:15]: do we need null check actions?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoResolverWithAddedModifiers.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+            _source = source;
+            _modifiers = modifiers;
+        }
+
+        public JsonTypeInfoResolverWithAddedModifiers WithAddedModifier(Action<JsonTypeInfo> modifier)
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoResolverWithAddedModifiers.cs:20]: null check?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoResolverWithAddedModifiers.cs (hunk 2 above)
Comment (krwq) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoResolverWithAddedModifiers.cs:15]: I see you've added null check in the public API

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoResolverWithAddedModifiers.cs (hunk 3 above)
Comment (krwq) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoResolverWithAddedModifiers.cs:20]: I see you've added null check in the public API so ignore this comment

**Review by krwq — APPROVED** [review]

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoResolverWithAddedModifiers.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoResolverWithAddedModifiers.cs:11]: We want the resolver instances to be immutable -- if we reused the same `List<T>` then calling `var bar = foo.WithAddedModifiers(...);` then `foo` would be mutated.

**Pull request merged**